engine/crates/engine/src/registry/resolvers/transformer.rs

```diff
@@ -275,66 +291,202 @@ impl Transformer {
-                                        serde_json::Value::String(STANDARD_NO_PAD.encode(string.as_bytes()));
+                                Value::Null => result.push(Value::Null),
+                                Value::String(ref string) => {
+                                    let new_value = Value::String(STANDARD_NO_PAD.encode(string.as_bytes()));
 
                                     result.push(new_value)
                                 }
                                 _ => return Err(Error::new("The resolved value is not a bytes string")),
                             }
                         }
 
-                        serde_json::Value::Array(result)
+                        Value::Array(result)
                     }
                     _ => return Err(Error::new("The resolved value is not a bytes string")),
                 };
 
-                Ok(ResolvedValue::new(new_value))
+                let mut new_value = ResolvedValue::new(new_value);
+                new_value.selection_data = resolved_value.selection_data.clone();
+
+                Ok(new_value)
             }
             Transformer::MongoTimestamp => {
                 let resolved_value =
                     last_resolver_value.ok_or_else(|| Error::new("Internal error resolving mongo timestamp"))?;
 
                 let value = match resolved_value.data_resolved() {
-                    serde_json::Value::Null => serde_json::Value::Null,
-                    serde_json::Value::Number(num) => serde_json::Value::Number(num.clone()),
-                    serde_json::Value::Object(object) => match object.get("T") {
-                        Some(serde_json::Value::Number(ms)) if ms.is_u64() => serde_json::Value::Number(ms.clone()),
+                    Value::Null => Value::Null,
+                    Value::Number(num) => Value::Number(num.clone()),
+                    Value::Object(object) => match object.get("T") {
+                        Some(Value::Number(ms)) if ms.is_u64() => Value::Number(ms.clone()),
                         _ => return Err(Error::new("Cannot coerce the initial value into a valid Timestamp")),
                     },
                     _ => return Err(Error::new("Cannot coerce the initial value into a valid Timestamp")),
                 };
 
-                Ok(ResolvedValue::new(value))
+                let mut new_value = ResolvedValue::new(value);
+                new_value.selection_data = resolved_value.selection_data.clone();
+
+                Ok(new_value)
+            }
+            Transformer::PostgresPageInfo => {
+                let resolved_value =
+                    last_resolver_value.ok_or_else(|| Error::new("Internal error resolving postgres page info"))?;
+
+                let mut rows: Vec<Value> = match resolved_value.data_resolved() {
+                    Value::Array(rows) => rows.clone(),
+                    _ => return Ok(resolved_value.clone()),
+                };
```

Comment:
I think you could make this clone less expensive in certain circumstances if you changed this function to take an owned `Option<ResolvedValue>` instead of an `Option<&ResolvedValue>` (which I have been intending to do at some point anyway) and then using take instead of data_resolved:

```suggestion
                if !resolved_value.data_resolved().is_array() {
                    return Ok(resolved_value)
                }
                let Value::Array(mut rows) = resolved_value.take() else {
                    unreachable!();
                }
```

`ResolvedValue::take` will _sometimes_ clone the underlying data, but it'll avoid it if it can.  At the very least the case where you're doing `FindMany.and_then(PostgresPageInfo)` should avoid the clone I think

Reply:
How much work would it be to do this ownership change in this PR? I've been thinking about doing so, but didn't want to _change all the things_ immediately.

Reply:
I don't think this is a large change, but it can definitely wait till you're not on holiday.